### DISCUSSION TITLE
[아이작의 황금고블린] Enum 이놈은 뭐 하는 놈일까??

### DISCUSSION THREAD

**Discussion by DevKang** [DeveloperAcademy-POSTECH/.github#69] [Topic]
---
```
// (*) 이 이벤트는 랜덤으로 발생합니다. 
// (**) 같은 컨셉으로 많이 많이 카피해주세요 (예: [예니의 황금 고블린] Protocol 시시콜콜하게 따져보자!)
```
---

문득 궁금해졌습니다. 

> Swift는 Enum활용이 꽃이라고 하는데... 우리는 얼마나 잘 알고 있나?

Swift에서 Enumeration이 중요하다는 것은 이미 많이 들어서 익숙하실거에요. 
이미 잘 사용하고 계신 분이 있다면 이 부분에 공감도 하실 것이라고 생각합니다. 
Swift의 Enum은 다른 언어보다 훨씬 풍부한 표현방식을 가지고 있고, 용법도 굉장히 다양합니다.
우선 깊이 있게 Enumeration에 대해서 공부해보기 이전에, **이 문법이 왜 존재하는지** 알아보면 좋지 않을까요?

댓글로 아래 사항들에 대해서 함께 이야기 해보면 좋을 것 같아요. 
**모든 항목에 대해서 다 답글로 달 필요는 없어요. 가볍게 토의 해봐요 🌞**

1. 내가 생각하는 Enumeration의 존재 이유는 무엇인가요?
2. Swift의 Enumeration을 특별하게 만드는 기능은 뭐라고 생각하세요?
3. 나는 Enum 이렇게 썼었다! 케이스를 코드로 공유해주세요 :-)

참고 리소스
[[Swift Docs] Enumerations](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/enumerations/#Matching-Enumeration-Values-with-a-Switch-Statement) (문서 최상단 도입부만 먼저 읽어보세요)

**Comment by DevKang** [comment]
저는 Enum은 [경우의 수]를 표현하기 위한 문법이라고 생각해요~! 
문자열을 표현하는 String, 정수를 표현하는 Int 처럼 어떤 개념이 특정한 경우의 수 안에서 표현을 해야할 때 Enum 타입을 사용한다고 이해하고 있습니다 🤠

**Comment by shippingpark** [comment]
안녕하세요! 3기 보노입니다. 다른 분들의 댓글 기다리며 제가 생각하는 enum에 대해 소소하게 댓글을 달아 봅니다...! 
(아마 2번 문항에 대한 댓글이지 않을까... 합니다!)

저는 enum을 굉장히 자주 쓰고... 또 좋아하는데요...! enum이 가진 가독성과 명확함을 좋아하기 때문입니다. 
개인적으로는 `String`으로 표현된 값을 `enum`으로 옮길 때 그러한 특성을 두드러지게 느끼는 것 같습니다. 

String과 비교하자면 enum은 아래의 장점을 가진다 생각합니다.
1. 명확한 표현을 가능하게 함
2. 허용되지 않은 값 방지 
3. 값타입의 메모리 효율 (String은 struct지만, 경우에 따라 힙에 올라감) 

그러니 저는 코드의 의미를 명확히 하고 잘못된 사용을 방지하기 위해 enum을 지향하며, 때때로는 메모리 관점에서 String보다 enum을 지향합니다. 

String을 사용할 지 enum을 사용할지 고민하는 제 개인적인 기준은 아래와 같습니다. 
1. 사용 경우를 제한할 수 있는가 
2. 케이스가 확장 될 여지가 없는가   

두 질문 다 `Yes`라면 가급적 enum을 선택하는 것 같습니다. 

+) String -> enum 으로 개선하는 - 메모리 효율과 관련된 이야기는 WWDC 2016 Understanding Swift Performance 영상을 통해 접하게 된 내용입니다! 
(*레퍼런스 카운팅 오버헤드를 개선하기 위해 String 보다 enum을 택하는 코드 예시가 존재합니다)

[개선 전]
![image](https://github.com/user-attachments/assets/88bcacfe-eab4-4cb4-93e8-4501442d3efa)


[개선 후]
![image](https://github.com/user-attachments/assets/24a41d11-70b7-48ca-bf62-682ba0027735)

**Comment by yeniful** [comment]
안녕하세요 1기 예니입니다. 

값들의 관리 측면에서 유용하다고 생각해요. 
같은 용도로 사용되는 여러개의 변수, 상수를 그냥 선언해서 사용하는 것 보다 
상위의 개념으로 묶고 Enum으로 만들어 주었을 때 가독성과 유지보수의 용이성이 높아진다고 생각합니다.
특히 서로 다른 A , B class/struct 안에서 비슷한 구조의 Enum을 사용하지만, 값을 다르게 선언하는 복잡한 상황에서 
혼란스러움을 줄여줄 수 있는 장치인 것 같아요.

무엇보다도 보노가 잘 설명해주신대로 허용되지 않는 값을 방지한다는게 Enum이 가장 빛나는 이유 같습니다-!
